### PR TITLE
P4 2615 bulk price updates

### PR DESCRIPTION
**Pull request by matt-crowson-moj**
Changes:

- This change introduces a new Spring shell command to support the generation of the following effective years prices for a supplier with the given multiplier. No current years prices are affected by this change it only ever creates new prices (or deletes and recreates new prices if run more than once.

**_Note: This will need auditing adding to it when available._**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/commands/ImportCommands.kt [hunk 1]
```diff
@@ -15,27 +16,42 @@ import java.time.temporal.ChronoUnit
 class ImportCommands(
   @Autowired private val importService: ImportService
 ) {
+
+  private val logger = LoggerFactory.getLogger(javaClass)
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/commands/ImportCommands.kt:20]: A small refactoring to add some logging.  No functional change to the underlying commands.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/config/ImporterConfiguration.kt [hunk 2]
```diff
@@ -27,5 +27,6 @@ class ImporterConfiguration {
   fun timeSource() = TimeSource { LocalDateTime.now(Clock.systemDefaultZone()) }
 
   @Bean
-  fun supplierPrices() = SupplierPrices { supplier -> priceRepository.findBySupplier(supplier) }
+  fun supplierPrices() =
+    SupplierPrices { supplier, year -> priceRepository.findBySupplierAndEffectiveYear(supplier, year) }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/config/ImporterConfiguration.kt:31]: This is used by the spreadsheet generation.  With the introduction of next years prices we now have to consider which effective prices are included in the generated spreadsheet i.e. we only include the prices we are interested in.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/config/SupplierPrices.kt [hunk 3]
```diff
@@ -4,5 +4,5 @@ import uk.gov.justice.digital.hmpps.pecs.jpc.price.Price
 import uk.gov.justice.digital.hmpps.pecs.jpc.price.Supplier
 
 fun interface SupplierPrices {
-  fun get(supplier: Supplier): List<Price>
+  fun get(supplier: Supplier, effectiveYear: Int): List<Price>
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/config/SupplierPrices.kt:7]: As above.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/BulkPricesService.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  fun addNextYearsPrices(supplier: Supplier, multiplier: Double) {
+    val now = timeSource.date()
+
+    logger.info("Adding $supplier prices for effective year ${nextEffectiveYearForDate(now)} with multiplier $multiplier")
+
+    priceRepository.deleteBySupplierAndEffectiveYear(supplier, nextEffectiveYearForDate(now))
+
+    priceRepository.flush()
+
+    val total = priceRepository.findBySupplierAndEffectiveYear(supplier, effectiveYearForDate(now)).map {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/service/BulkPricesService.kt:35]: This essential does the following taken from [Confluence](https://dsdmoj.atlassian.net/wiki/spaces/PECS4/pages/2649948327/How+to+update+JPC+effective+date) page:

```
insert into prices(price_id, added_at, price_in_pence, supplier, 
from_location_id, to_location_id, effective_year)
  select uuid_generate_v4(), now(), price_in_pence * 2 as price_in_pence,
  supplier, from_location_id, to_location_id, 2021
    from prices
    where effective_year = 2020
```

File: src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/spreadsheet/PricesSpreadsheetGenerator.kt [hunk 5]
```diff
@@ -36,23 +37,33 @@ class PricesSpreadsheetGenerator(
...
 
       SummarySheet(workbook, header).also { logger.info("Adding summaries.") }.apply { writeSummaries(summaries) }
 
-      SupplierPricesSheet(workbook, header).also { logger.info("Adding prices used for supplier $supplier.") }.apply { writePrices(supplierPrices.get(supplier)) }
+      SupplierPricesSheet(
+        workbook,
+        header
+      ).also {
+        logger.info("Adding $supplier prices for effective year ${effectiveYearForDate(startDate)}.")
+      }.apply { writePrices(supplierPrices.get(supplier, effectiveYearForDate(startDate))) }
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/pecs/jpc/spreadsheet/PricesSpreadsheetGenerator.kt:66]: Ensures we are using the correct effective years prices when putting into generated spreadsheet.

**Review by Thource — APPROVED** [review]

**Review by Thource — APPROVED** [review]

**Pull request merged**